src/student_success_tool/analysis/pdp/features/student_term.py

```diff
@@ -382,24 +370,62 @@ def sum_val_equal_cols_by_group(
     )
 
 
-def multicol_aggs_by_group(
+def _rename_sum_by_group_col(col: str) -> str:
+    return f"{constants.NUM_COURSE_FEATURE_COL_PREFIX}_{col}"
+
+
+def multicol_grade_aggs_by_group(
     df: pd.DataFrame,
     *,
     grp_cols: list[str],
-    grade_col: str = "course_grade_numeric",
-    section_grade_col: str = "section_course_grade_numeric_mean",
+    grade_col: str = "grade",
+    grade_numeric_col: str = "course_grade_numeric",
+    section_grade_numeric_col: str = "section_course_grade_numeric_mean",
 ) -> pd.DataFrame:
-    df_grped = df.groupby(by=grp_cols, observed=True, as_index=False)
-    df_applied = df_grped.apply(
-        num_courses_grade_above_section_avg,
-        grade_col=grade_col,
-        section_grade_col=section_grade_col,
-        include_groups=False,
+    return (
+        df.loc[:, grp_cols + [grade_col, grade_numeric_col, section_grade_numeric_col]]
+        # compute intermediate column values all at once, which is efficient
+        .assign(
+            course_grade_is_failing_or_withdrawal=ft.partial(
+                _course_grade_is_failing_or_withdrawal,
+                grade_col=grade_col,
+                grade_numeric_col=grade_numeric_col,
+            ),
+            course_grade_above_section_avg=ft.partial(
+                _course_grade_above_section_avg,
+                grade_numeric_col=grade_numeric_col,
+                section_grade_numeric_col=section_grade_numeric_col,
+            ),
+        )
+        .groupby(by=grp_cols, observed=True, as_index=False)
+        # so that we can efficiently aggregate those intermediate values per group
+        .agg(
+            num_courses_grade_is_failing_or_withdrawal=(
+                "course_grade_is_failing_or_withdrawal",
+                "sum",
+            ),
+            num_courses_grade_above_section_avg=(
+                "course_grade_above_section_avg",
+                "sum",
+            ),
+        )
     )
-    assert isinstance(df_applied, pd.DataFrame)  # type guard
-    # pandas does not give a shit about type guards, ignoring the complaint below
-    return df_applied.rename(columns={None: "num_courses_grade_above_section_avg"})  # type: ignore
 
 
-def _rename_sum_by_group_col(col: str) -> str:
-    return f"{constants.NUM_COURSE_FEATURE_COL_PREFIX}_{col}"
+def _course_grade_is_failing_or_withdrawal(
+    df: pd.DataFrame,
+    grade_col: str = "grade",
+    grade_numeric_col: str = "course_grade_numeric",
+) -> pd.Series:
+    return (
+        df[grade_col].isin({"F", "W"})
+        | df[grade_numeric_col].between(0.0, 2.0, inclusive="left")
```

Comment:
I think this would include C- per [these grade conversions](https://bigfuture.collegeboard.org/plan-for-college/get-started/how-to-convert-gpa-4.0-scale). Although of course hard to say what PDP grades might look like, want to use 1.7 instead?

Reply:
It does not: the bound is left-inclusive, so strictly < 2.0

Reply:
Wait, what? A C- = 1.7 ??

Reply:
k I will fix this by passing through the `min_passing_grade` arg, which is how I should've done it originally

Reply:
thanks for catching this btw! apologies for the scattered comments -- trying to finish modeling/bias assessment and the presentation, and i'm worn out 😮‍💨

Reply:
All good!! I definitely understand being worn out.